### ISSUE TITLE
Capture identity attributes (enduser.*) for applications using Spring Security

### ISSUE THREAD

**Issue by philsttr** [open-telemetry/opentelemetry-java-instrumentation#9400]
### Is your feature request related to a problem? Please describe.

[General identity attributes](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/attributes.md#general-identity-attributes) (`enduser.id`, `enduser.role`, `enduser.scope`) are not captured for SERVER spans in applications using Spring Security OAuth2 Resource Server for [Servlet](https://docs.spring.io/spring-security/reference/servlet/oauth2/resource-server/index.html) or [WebFlux](https://docs.spring.io/spring-security/reference/reactive/oauth2/resource-server/index.html) based applications.

The existing servlet instrumentation currently captures the `enduser.id` in a couple places ([here](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/44e908e52cd9efa00b4df897ed95c95cc66b04ed/instrumentation/servlet/servlet-common/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/servlet/ServletAdditionalAttributesExtractor.java#L50) and [here](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/44e908e52cd9efa00b4df897ed95c95cc66b04ed/instrumentation/servlet/servlet-common/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/servlet/BaseServletHelper.java#L142)).  However, in an Servlet-based application using Spring Security OAuth2 Resource Server, the `request.getUserPrincipal()` returns null at this level.  It returns null here because Spring Security wraps that request object in [a request wrapper](https://docs.spring.io/spring-security/site/docs/current/api/org/springframework/security/web/servletapi/SecurityContextHolderAwareRequestWrapper.html) at a higher level, and the user principal is only returned from the wrapper at the higher level.  In other words, since the servlet instrumentation only has access to the lower level request, and not the higher level request wrapper, the user principal seen by the servlet instrumentation is null, and therefore the `enduser.id` is not captured.

In addition, there is no existing instrumentation for capturing the `enduser.id` in WebFlux applications.

### Describe the solution you'd like

I would like [general identity attributes](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/attributes.md#general-identity-attributes) (`enduser.id`, `enduser.role`, `enduser.scope`) to be captured for SERVER spans in applications using Spring Security OAuth2 Resource Server for [Servlet](https://docs.spring.io/spring-security/reference/servlet/oauth2/resource-server/index.html) and [WebFlux](https://docs.spring.io/spring-security/reference/reactive/oauth2/resource-server/index.html) based applications.

### Additional context

While the new instrumentation probably needs to be Spring Security specific, it does not necessarily need to be specific to OAuth2 Resource Servers authentication/authorization.  i.e. The new instrumentation could probably work for any type of Spring Security based authentication/authorization.

**Comment by mateuszrzeszutek** [comment]
Hey @philsttr ,

I think this'd be a welcome addition 👍 
We'd have to make the new instrumentation disabled by default because of the
>Given the sensitive nature of this information, SDKs and exporters SHOULD drop these attributes by default and then provide a configuration parameter to turn on retention for use cases where the information is required and would not violate any policies or regulations.

fragment in the spec theough.

**Comment by philsttr** [comment]
Agreed.  Probably should do that for the existing servlet instrumentation as well.

**Comment by swar8080** [comment]
Maybe these attributes could be captured for most spring security use cases by instrumenting [SecurityContextHolder#setContext](https://docs.spring.io/spring-security/site/docs/4.0.x/apidocs/org/springframework/security/core/context/SecurityContextHolder.html)? From there a list of [GrantedAuthority](https://docs.spring.io/spring-security/site/docs/4.0.x/apidocs/org/springframework/security/core/GrantedAuthority.html) are accessible, which is where roles and scopes are stored. The spring convention is that [roles](https://docs.spring.io/spring-security/reference/5.7/servlet/authorization/expression-based.html) start with the prefix `ROLE_`, and the oauth2 resource server library [stores scopes as granted authorities](https://github.com/spring-projects/spring-security/blob/main/oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/authentication/JwtGrantedAuthoritiesConverter.java#L46) prefixed with `SCOPE_` by default

One of the things i'm not sure about is this part of the convention:
> It is expected this information would be propagated unchanged from node-to-node within the system using the Baggage mechanism. These attributes should not be used to record system-to-system authentication attributes.

For example, i've seen production queue consumer code that manually calls `SecurityContextHolder#setContext`. Including these attributes as baggage on HTTP requests by that queue consumer could be unexpected. So maybe more focused instrumentation is needed to avoid that, like on the specific spring security oauth2 resource server http filters.

**Comment by philsttr** [comment]
Filed #9740 for updating the existing servlet instrumentation to not capture `enduser.id` by default.

**Comment by philsttr** [comment]
I'm unsure how this requirement would be best implemented...

> Given the sensitive nature of this information, SDKs and exporters SHOULD drop these attributes by default and then provide a configuration parameter to turn on retention for use cases where the information is required and would not violate any policies or regulations.

Should this requirement be implemented by:
* having each instrumentation that captures the `endpoint.*` attributes look at a configuration value to determine if it should capture them, OR
* have each instrumentation _always_ capture them, but then have something later (the exporter?) filter them out based on configuration
?

**Comment by mateuszrzeszutek** [comment]
IMO option 1 should be chosen as the solution here -- treat this attribute as an opt-in thing, and only collect it when the user explicitly configures the agent to do so.

> If we go with option 1 (where instrumentations check a config property), then perhaps `otel.instrumentation.common.capture-enduser.enabled=true/false`

SGTM 👍

**Comment by trask** [comment]
> IMO option 1 should be chosen as the solution here -- treat this attribute as an opt-in thing, and only collect it when the user explicitly configures the agent to do so.

related: https://github.com/open-telemetry/semantic-conventions/issues/128#issuecomment-1775594259

> The guidance to not generate PII by default is temporary, pending better enable/disable features in the SDK for full user control.

**Comment by philsttr** [comment]
I just filed https://github.com/spring-projects/spring-security/issues/14046 to see if the Spring Security team would consider adding a hook that could be used by instrumentation to capture these attributes, or to see if they would recommend a different approach.

**Comment by philsttr** [comment]
New approach, as suggested by the spring-security folks...

Add a filter into the spring security filter chain.  The new filter is ordered after the filters that setup the spring security context.  Therefore, the new filter can retrieve the Authentication and populate the enduser.* attributes.

More specifically

* Servlet apps: instrument `org.springframework.security.config.annotation.web.builders.HttpSecurity.build()` to add a filter the `AuthorizationFilter`
* Reactive apps: instrument `org.springframework.security.config.web.server.ServerHttpSecurity.build()` to add a filter before the logout filter

**Comment by philsttr** [comment]
For reference:  https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/9777 adds the OpenTelemetry instrumentation described above

**Comment by jaydeluca** [comment]
Closing this as resolved via [#9777](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/9777). 

If there is anything outstanding feel free to re-open or open a new issue

**Comment by trask** [comment]
just a heads up for anyone here on this issue
* #13203
